core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -359,11 +380,38 @@ final class ZSTM[-R, +E, +A] private[stm] (
    */
   def fold[B](f: E => B, g: A => B)(implicit ev: CanFail[E]): URSTM[R, B] =
     self.continueWithM {
-      case TExit.Fail(e)    => ZSTM.succeedNow(f(e))
+      case TExit.Fail(c)    => c.failureOrCause.fold(e => ZSTM.succeedNow(f(e)), ZSTM.halt(_))
+      case TExit.Succeed(a) => ZSTM.succeedNow(g(a))
+      case TExit.Retry      => ZSTM.retry
+    }
+
+  /**
+   * A more powerful version of `fold` that allows recovering from any kind of failure.
+   */
+  def foldCause[B](f: Cause[E] => B, g: A => B)(implicit ev: CanFail[E]): URSTM[R, B] =
+    self.continueWithM {
+      case TExit.Fail(c)    => ZSTM.succeedNow(f(c))
       case TExit.Succeed(a) => ZSTM.succeedNow(g(a))
       case TExit.Retry      => ZSTM.retry
     }
 
+  /**
+   * A more powerful version of `foldM` that allows recovering from any kind of failure.
+   *
+   * todo: is this actually useful?
+   */
+  def foldCauseM[R1 <: R, E2, B](
```

Comment:
Seems perfectly useful. Might be worth adding a note that since these are transactions versus effects the `Cause` could contain multiple parallel failures but not interruption or die failures. And I believe at least as of today there are not situations in which the cause would have multiple sequential failures.